fix(aws): Update diffing logic for AWS upsertSecurityGroupOperation

aravindmd opened the pull request:
- Update the logic to find the delta that needs to be removed / added and updated during security group upsert operation 
- Since we now have description included in the filtered list of rules , we need to provide a way to update description when required.
- Updates to rules are done using aws `updateSecurityGroupRuleDescriptionsIngress` api

marchello2000 commented on clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/securitygroup/SecurityGroupIngressConverter.groovy:
```diff
@@ -130,4 +127,48 @@ class SecurityGroupIngressConverter {
+    List<IpPermission> tobeAdded = new ArrayList<>()
+    List<IpPermission> tobeRemoved = new ArrayList<>()
+    List<IpPermission> tobeUpdated = new ArrayList<>()
+    List<IpPermission> filteredList = newList.findAll { elements -> elements.userIdGroupPairs.size() == 0 }
```
nit: I prefer to either use something completely non-descriptive like the groovy `it` or (preferably) something descriptive like `ipPermission` 
(and same below)
(also since you have this nice convention already... `filteredNewList`)
```suggestion
    List<IpPermission> filteredNewList = newList.findAll { ipPermission -> ipPermission.userIdGroupPairs.size() == 0 }
```

marchello2000 commented on clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/securitygroup/SecurityGroupIngressConverter.groovy:
```diff
@@ -130,4 +127,48 @@ class SecurityGroupIngressConverter {
+    filteredList.forEach({ newListEntry ->
+      IpPermission match = findIpPermission(filteredExisistingRuleList, newListEntry)
+      if (match) {
+        if (newListEntry.ipv4Ranges.collect { it.description }.find() != null
```
i think this would read better if you had (for both of these obv but github won't let me do that)
```suggestion
        if (newListEntry.ipv4Ranges.collect { it.description }.any()
```

all you care is there is at least one description, right?

marchello2000 commented on clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/securitygroup/SecurityGroupIngressConverter.groovy:
```diff
@@ -130,4 +127,48 @@ class SecurityGroupIngressConverter {
+  }
+
+  static IpPermission findIpPermission(List<IpPermission> existingList, IpPermission ipPermission) {
+    existingList.find { it ->
```
why not rely on the [`IpPermission.equals`](https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-ec2/src/main/java/com/amazonaws/services/ec2/model/IpPermission.java#L697) method I think it does exactly what you want (but please double check)

aravindmd replied:
It does not exactly do what is required, i will leave this logic in , unless you feel we need to update it

marchello2000 commented on clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/securitygroup/UpsertSecurityGroupAtomicOperation.groovy:
```diff
@@ -101,8 +101,36 @@ class UpsertSecurityGroupAtomicOperation implements AtomicOperation<Void> {
-    List<IpPermission> ipPermissionsToRemove = existingIpPermissions - ipPermissionsFromDescription.converted
+    Map ipPermissionMap = SecurityGroupIngressConverter.computeLatestIpRules(ipPermissionsFromDescription.converted, existingIpPermissions)
+
+    List<IpPermission> ipPermissionsToAdd = ipPermissionMap.get("tobeAdded")
```
nit:
```suggestion
    List<IpPermission> ipPermissionsToAdd = ipPermissionMap.["tobeAdded"]
```

marchello2000 commented on clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/securitygroup/SecurityGroupIngressConverter.groovy:
```diff
@@ -130,4 +127,48 @@ class SecurityGroupIngressConverter {
     }.flatten().unique()
   }
+
+  static Map<String, List> computeLatestIpRules(List<IpPermission> newList, List<IpPermission> existingRules) {
```
maybe 
```suggestion
  static Map<String, List> computeIpRuleDelta(List<IpPermission> newList, List<IpPermission> existingRules) {
```

marchello2000 commented on clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/securitygroup/SecurityGroupLookupFactory.groovy:
```diff
@@ -249,6 +250,14 @@ class SecurityGroupLookupFactory {
+        groupId: securityGroup.groupId,
+        ipPermissions: ipPermissionsToUpdate
+      ))
+      securityGroup.ipPermissions.addAll(ipPermissionsToUpdate)
```
is this right? since it's just a list, won't this mean that an updated iprule will show up twice? once with old description and once with new?
I am not sure it's used today, but even if not, in the future someone might assume it works differently

aravindmd replied:
I have removed it , not sure if its used today

aravindmd commented:
> A few nits, and a few questions. Also, probably worth adding a unittest that the differ works as expected.
> and definitely worth explaining why the "fancy" differ exists in a comment in one of these classes.
> 
> Otherwise, looks to me like the logic makes sense

Added more comments about the diff logic and added UT's for the same

marchello2000 commented on clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/securitygroup/SecurityGroupIngressConverter.groovy:
```diff
@@ -130,4 +127,59 @@ class SecurityGroupIngressConverter {
+    List<IpPermission> tobeRemoved = new ArrayList<>()
+    List<IpPermission> tobeUpdated = new ArrayList<>()
+    List<IpPermission> filteredNewList = newList.findAll { ipPermission -> ipPermission.userIdGroupPairs.size() == 0 }
+    List<IpPermission> filteredExistingRuleList = existingRules.findAll { elements -> elements.userIdGroupPairs.size() == 0 }
```
nit
```suggestion
    List<IpPermission> filteredExistingRuleList = existingRules.findAll { ipPermission -> ipPermission.userIdGroupPairs.size() == 0 }
```

cfieber replied:
```suggestion
    List<IpPermission> filteredExistingRuleList = existingRules.findAll { ipPermission -> ipPermission.userIdGroupPairs.isEmpty() }
```

cfieber commented on clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/securitygroup/SecurityGroupIngressConverter.groovy:
```diff
@@ -130,4 +127,59 @@ class SecurityGroupIngressConverter {
+    List<IpPermission> tobeAdded = new ArrayList<>()
+    List<IpPermission> tobeRemoved = new ArrayList<>()
+    List<IpPermission> tobeUpdated = new ArrayList<>()
+    List<IpPermission> filteredNewList = newList.findAll { ipPermission -> ipPermission.userIdGroupPairs.size() == 0 }
```
```suggestion
    List<IpPermission> filteredNewList = newList.findAll { ipPermission -> ipPermission.userIdGroupPairs.isEmpty() }
```

cfieber commented on clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/securitygroup/SecurityGroupIngressConverter.groovy:
```diff
@@ -130,4 +127,59 @@ class SecurityGroupIngressConverter {
+   * - If a new rule has description value add it to update list to make it consistent.
+   * - If new rule has no description value set, ignore.
+   */
+  static Map<String, List> computeIpRuleDelta(List<IpPermission> newList, List<IpPermission> existingRules) {
```
having this return Map<String, List> feels a bit brittle
I'd suggest making a concrete return type object 
```
@Canonical
static class IpRuleDelta {
  List<IpPermission> toAdd
  List<IpPermission> toRemove
  List<IpPermission> toUpdate
}
```

aravindmd replied:
good suggestion , updated to use a concrete object.

cfieber commented on clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/securitygroup/UpsertSecurityGroupAtomicOperationUnitSpec.groovy:
```diff
@@ -428,4 +430,133 @@ class UpsertSecurityGroupAtomicOperationUnitSpec extends Specification {
 
   }
 
+  void "should update ingress and add by name for missing ingress security group in EC2 classic"() {
```
thanks for adding tests!

aravindmd commented on clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/securitygroup/SecurityGroupIngressConverter.groovy:
```diff
@@ -130,4 +127,65 @@ class SecurityGroupIngressConverter {
+      (((it.ipv4Ranges.collect { it.cidrIp }.sort() == ipPermission.ipv4Ranges.collect { it.cidrIp }.sort()
+        && it.fromPort == ipPermission.fromPort
+        && it.toPort == ipPermission.toPort
+        && it.ipProtocol == ipPermission.ipProtocol) && !ipPermission.ipv4Ranges.isEmpty())
```
Added this to eliminate cases where it would return a match for empty ranges
